### DISCUSSION TITLE
C# Language Ideas from Planning Domain Modeling

### DISCUSSION THREAD

**Discussion by AdamSobieski** [dotnet/csharplang#3071] [Language Ideas]
## Introduction
I am working on a .NET library with which to develop [automated planning](https://en.wikipedia.org/wiki/Automated_planning_and_scheduling) software. In the course of the project, I was creating an object model to parse [PDDL](https://en.wikipedia.org/wiki/Planning_Domain_Definition_Language) planning domains and problems into. Planning domains include: types, constants, predicates, numeric fluents, object fluents, constraints, rules and actions.

While working on that object model for planning domains and problems, I started creating a new modeling / programming language. The new language is based on C#. It is a work in progress.

I'd like to share a rough-draft sketch of the new language in hopes that it might inspire some new C# and .NET features.

## Rough-draft Sketch
```
using OtherNamespace1;
using OtherNamespace2;

namespace Example
{
    [Metadata]
    public class Type1;
    public class Type2;
    public class Type3 : Type1;

    [Metadata]
    public predicate bool F1(Type1 p1, Type2 p2);
    public predicate bool F2(Type1 p1, Type2 p2, Type3 p3)
    {
        get
        {
        }
        set
        {
        }
    }
    public predicate bool F3(Type1|Type2 p1); // intersection and union types

    [Metadata]
    public predicate number F4(Type1 p1, Type2 p2); // numeric fluents

    [Metadata]
    public predicate Type3 F5(Type1 p1, Type2 p2); // object fluents
  
    [Metadata]
    public Type1 Atom1; // constants

    [Metadata]
    public constraint Constraint1 // constraints
    {
    }

    [Metadata]
    public derivation Rule1 // rules
    {
    }

    [Metadata]
    public action Example1(Type1 p1, Type2 p2) // actions
    {
        preconditions
        {
            F1(p1, p2);
        }
        effects
        {
      
        }
        body
        {
            Example4(p1);
            Example5(p1, p2);
        }
    }

    [AbstractionOf(typeof(Example1))] // example metadata scenario
    public action Example2(Type1 p1, Type2 p2)
    {
        preconditions
        {
            
        }
        effects
        {
      
        }
        body
        {

        }
    }

    [Metadata]
    public durative action Example3(Type1 p1, Type2 p2) // durative actions
    {
        conditions
        {

        }
        effects
        {
      
        }
        duration
        {

        }
        body
        {

        }
    }
}
```

## Related Issues
https://github.com/dotnet/csharplang/issues/399 - Discussion: Union and Intersection types

**Comment by iam3yal** [comment]
What are the semantics of these "features"? how do they relate? what's the benefit of having them? why do we need them? all I see is text.

**Comment by AdamSobieski** [comment]
Planning domains include more than types, interfaces and delegates. The semantics include first-class:

* types
* constants
* predicates
* numeric fluents
* object fluents
* constraints
* rules
* actions

## Predicates, Numeric Fluents and Object Fluents
Predicates, numeric fluents and object fluents can be used in a number of scenarios including, but not limited to, .NET logic programming.

In the syntax, above, one can see the getter and setter on predicate "F2". With this feature, custom functionality can be provided for obtaining a value from or assigning a value to the predicate, fluent or object fluent. This is referred to in publications as semantic attachments [1][2][3].

## Actions
Looking at action definitions, C# language ideas include adding more optional sections to method definitions.

One idea for so doing involves code contracts. In issue https://github.com/dotnet/roslyn/issues/119 , C# syntactic expressiveness for `requires` and `ensures`  was discussed.

```
public int foo(string x)
{
  requires { x != null; }
  body
  {
    return x.Length;
  }
  ensures { return = x.Length; }
}
```

## References
[1] Dornhege, Christian, Patrick Eyerich, Thomas Keller, Sebastian Trüg, Michael Brenner, and Bernhard Nebel. "Semantic attachments for domain-independent planning systems." In Nineteenth International Conference on Automated Planning and Scheduling. 2009.

[2] Hertle, Andreas, Christian Dornhege, Thomas Keller, and Bernhard Nebel. "Planning with Semantic Attachments: An Object-Oriented View." In ECAI, vol. 242, pp. 402-407. 2012.

[3] Katz, Michael, Dany Moshkovich, and Erez Karpas. "Semi-Black Box: Rapid Development of Planning Based Solutions." In Thirty-Second AAAI Conference on Artificial Intelligence. 2018.

**Comment by HaloFour** [comment]
I don't think pointing the issue to a bunch of papers written on the subject are going to help.  We need concrete examples of where all of this syntax will be used to solve problems in C# programs, why existing syntax is not sufficient to solve those problems and how this new syntax addresses those short comings.  You can start with an explanation as to how and why this proposal is different and better than method contracts.  This proposal looks much more complicated than method contracts which despite having a champion and being discussed for several years seems to be no closer to actually being implemented as a language feature.

**Comment by AdamSobieski** [comment]
Method contracts are presently being discussed in issue https://github.com/dotnet/csharplang/issues/105 .

I prefer the syntax:

```
public int foo(string x)
{
  requires { x != null; }
  body
  {
    return x.Length;
  }
  ensures { return = x.Length; }
}
```

to the syntax:

```
public int foo(string x)
  requires x != null
  ensures return = x.Length
{
    return x.Length;
}
```

The two can be said to map to one another, however.

One concern with respect to some of the method contract syntax utilized in ongoing discussions is readability when defining lambda expressions. I think that this is more readable:

```
var lambda = (string x) => {
  requires { x != null; }
  body { return x.Length; }
  ensures { return = x.Length; }
};
```

**Comment by CyrusNajmabadi** [comment]
i'm not sure what the purpose of this issue is.  Is this a proposal for changes to c#?

> I think that this is more readable:

I think concerns about the syntax for contracts should be given to those proposals.  Note that syntax is usually the last thing figured out for complex language topics

**Comment by AdamSobieski** [comment]
I was hoping to open a discussion issue rather than a specific proposal issue. Perhaps a variety of new C# programming language features can be thought of when considering the modeling of planning domains. The method contracts syntax topic was an example.

**Comment by CyrusNajmabadi** [comment]
What is a planning domain?  Why would it benefit from language changes instead of just apis?

**Comment by AdamSobieski** [comment]
## Example Planning Domain and Problem
Here is a [STRIPS](https://en.wikipedia.org/wiki/Stanford_Research_Institute_Problem_Solver) example of a planning domain definition [2]:

```
(define (domain gripper-strips)
  (:predicates (room ?r) (ball ?b) (gripper ?g) (at-robby ?r)
               (at ?b ?r) (free ?g) (carry ?o ?g))
  (:action move
   :parameters (?from ?to)
   :precondition (and (room ?from)
                      (room ?to)
                      (at-robby ?from))
   :effect (and (at-robby ?to)
                (not (at-robby ?from))))
  (:action pick
   :parameters (?obj ?room ?gripper)
   :precondition (and (ball ?obj)
                      (room ?room)
                      (gripper ?gripper)
                      (at ?obj ?room)
                      (at-robby ?room)
                      (free ?gripper))
   :effect (and (carry ?obj ?gripper)
                (not (at ?obj ?room))
                (not (free ?gripper))))
  (:action drop
   :parameters (?obj ?room ?gripper)
   :precondition (and (ball ?obj)
                      (room ?room)
                      (gripper ?gripper)
                      (carry ?obj ?gripper)
                      (at-robby ?room))
   :effect (and (at ?obj ?room)
                (free ?gripper)
                (not (carry ?obj ?gripper)))))
```

With a domain, we can define a problem. Here is an example of a problem definition:

```
(define (problem strips-gripper2)
    (:domain gripper-strips)
    (:objects rooma roomb ball1 ball2 left right)
    (:init (room rooma)
           (room roomb)
           (ball ball1)
           (ball ball2)
           (gripper left)
           (gripper right)
           (at-robby rooma)
           (free left)
           (free right)
           (at ball1 rooma)
           (at ball2 rooma))
    (:goal (at ball1 roomb)))
```

The example _domain definition_ describes or models a world with: robots, grippers, balls, rooms. It defines three actions: move, pick and drop. The gist is that robots can be in rooms, move between rooms, pick things and and drop things with their grippers.

The example _problem definition_ describes a specific initial situation: there are two rooms (rooma, roomb), two balls (ball1,  ball2), two grippers (left, right), the robot is in a room (rooma), the grippers are empty, ball1 is in rooma, ball2 is in roomb.

The example problem definition also describes a goal: that a specific ball (ball1) be in a specific room (roomb).

The planner processes the domain and the problem and outputs a plan. In the example, the output involves a robot moving around, using its grippers... picking up a ball, moving to a room, dropping a ball.

The example domain and problem, while illustrative, don't showcase the newest features of [PDDL](https://en.wikipedia.org/wiki/Planning_Domain_Definition_Language). A concise history of the PDDL language (versions 1.2, 2.1, 2.2, 3.0 and 3.1) is available here: https://planning.wiki/ref . There also exist various [successors / variants / extensions to PDDL](https://en.wikipedia.org/wiki/Planning_Domain_Definition_Language#Successors/variants/extensions_of_PDDL).

A number of PDDL examples are available at: https://github.com/potassco/pddl-instances .

## PDDL and Visual Studio Code
Interestingly, there is language support for PDDL in Visual Studio Code. The project by Jan Dolejsi is available here: https://marketplace.visualstudio.com/items?itemName=jan-dolejsi.pddl .

One can define domains and problems in Visual Studio Code and easily connect to planner services which, at the push of a button, provide solutions or plans.

The default such service is http://planning.domains which also allows developers to author planning domains and problems on its website: http://editor.planning.domains/ .

## Making Planning Domain Modeling More Approachable
Some opine that planning domain modeling could be made more approachable. There is a learning curve to using PDDL and users intuitively understand other modern programming languages such as C#.

Regardless of the language used, preconditions and effects are described as the most difficult aspect of writing planning domains for new users [1].

## .NET
Were a new or existing .NET language to support planning domain modeling scenarios, there would be to consider first-class: types, constants, predicates, numeric fluents, object fluents, constraints, rules and actions.

Some of these constructs could convenience or facilitate other .NET languages, e.g. logic programming languages. Adding the aforementioned constructs could be CLI/CLR topics or class library topics.

PDDL supports union types which is why issue https://github.com/dotnet/csharplang/issues/399 is relevant.

Beyond parsing PDDL (and/or a new language) into a useful runtime model, one can ask: could a planning domain be comprised of contents from multiple files or modules? Could planning domain contents "compile to" or be stored in modular, reusable .NET assemblies?

## References
[1] Katz, Michael, Dany Moshkovich, and Erez Karpas. "Semi-Black Box: Rapid Development of Planning Based Solutions." In Thirty-Second AAAI Conference on Artificial Intelligence. 2018.
[2] Veloso, Manuela. "PDDL by Example." Carnegie Mellon University (2015).

**Comment by orthoxerox** [comment]
It looks to me tthat planning domain modeling would be better served by creating a DSL that compiles to CIL or C# rather than extension of C# itself. If you think that a specific feature could have broad applicability (like union types you've mentioned), why not open a separate issue for it?

**Comment by CyrusNajmabadi** [comment]
I'm not really seeing what this has to do with C#.  Are tehre specific changes you are asking for in particular?  Can you distill this down at all?

**Comment by AdamSobieski** [comment]
The above language in the rough-draft sketch can be described as a C#-based DSL. When one bases a DSL on C#, in addition to the DSL being approachable to most software developers, one can refer to existing C# projects and files when eventually working on code editor and IDE integration (e.g. https://github.com/dotnet/csharp-tmLanguage for syntax highlighting).

Those are good recommendations with respect to compiling the new DSL to CIL or transpiling it to C#. Presently, in the aforementioned software library project, PDDL and new DSL both parse into the same object model which includes its own custom type system. Transpiling from PDDL to the new DSL should be readily implementable.

As PDDL and the new DSL include union types, issue https://github.com/dotnet/csharplang/issues/399 is relevant. That is, if .NET supported intersection and union types, the new DSL (as well as a number of other languages) could more readily compile to CIL.

Distilling, some specific ideas for C#/.NET:
1. First-class Boolean predicates, numeric fluents, and object fluents. Developers can choose to specify getters and setters for these.
2. First-class constraints and rules (derivations).
3. Describing the preconditions and effects of actions or methods.

**Comment by CyrusNajmabadi** [comment]
> First-class Boolean predicates, numeric fluents, and object fluents. 

What are these things?

**Comment by AdamSobieski** [comment]
## Boolean Predicates

In a strongly-typed logical model, a _predicate_ can be modeled as having zero or more typed parameters and, when provided with arguments, results in a compound expression.

Let `P` be a Boolean predicate. Let `x`, `y` and `z` be arguments, objects with types. Then we can refer to `P(x, y, z)` as a compound expression. As `P` is a Boolean predicate, compound expressions involving `P` are either true or false for every valid triple of arguments.

For example, we can define `Carrying` as a Boolean predicate, something like:

```
public predicate bool Carrying(Robot robot, Ball ball);
```

`Carrying(r, b)` can be described as true if Ball `b` is carried by Robot `r` and false otherwise.

Depending upon how we want to address time, we might also choose to define `Carrying` in this way:

```
public predicate bool Carrying(Robot robot, Ball ball, DateTime time);
```

## Numeric Fluents

A _numeric fluent_ is where the compound expression is, instead of Boolean, numeric. For example, `AmountOfBattery(r)` could be an expression which is of a numeric variety, the amount of battery charge that a robot `r` has.

```
public predicate float AmountOfBattery(Robot robot);
```

## Object Fluents

An _object fluent_ is where the compound expression is, beyond Boolean or numeric, of an arbitrary type from the domain or model.

## Adding Custom Getters and Setters

Next, we can add getters and setters to indicate that a particular predicate provides custom procedures for getting the value of a compound expression or assigning a value to it. This is based on “semantic attachments” [1][2].

```
public predicate bool Carrying(Robot robot, Ball ball, DateTime time)
{
  get { return Database.Query(robot, ball, time); }
  set { Database.Store(robot, ball, time, value); }
}
```

There should be a means of efficiently checking whether a predicate defines custom procedures for its getter and setter.

## Related Topic: Multidimensional Properties

A somewhat related topic could be multidimensional properties: https://docs.microsoft.com/en-us/cpp/dotnet/how-to-use-properties-in-cpp-cli?view=vs-2019#multidimensional-properties .

Using multidimensional properties, a C++/CLI example:
```
world->Carrying[robot, ball, time] = true;
```

or, using static multidimensional properties:

```
World::Carrying[robot, ball, time] = true;
```

## Terminology

The terminology “compound” comes from [Prolog syntax and semantics](https://en.wikipedia.org/wiki/Prolog_syntax_and_semantics).

The terminology “numeric fluent” and “object fluent” are from the [history of the PDDL language](https://planning.wiki/ref).

The terminology “semantic attachments” comes from Weyhrauch [1] to describe the attachment of an interpretation to a predicate symbol using an external procedure.

## References

[1] Weyhrauch, Richard W. "Prolegomena to a theory of mechanized formal reasoning." Artificial intelligence 13, no. 1-2 (1980): 133-170.
[2] Dornhege, Christian, Patrick Eyerich, Thomas Keller, Sebastian Trüg, Michael Brenner, and Bernhard Nebel. "Semantic attachments for domain-independent planning systems." In Nineteenth International Conference on Automated Planning and Scheduling. 2009.

**Comment by CyrusNajmabadi** [comment]
I don't really see why you would need anything special here as this looks to be basically identical to what .net and c# already have with delegates

**Comment by AdamSobieski** [comment]
Predicates, numeric fluents and object fluents more closely resemble properties than either methods or delegates as one can both get values from and assign values to compound expressions.

When one defines custom getters and setters, this makes sense:

```
bool b = P(x, y, z);
```

but so does this:

```
P(x, y, z) = true;
```

With respect to ideas for C#, we can add [_multidimensional properties_](https://docs.microsoft.com/en-us/cpp/dotnet/how-to-use-properties-in-cpp-cli?view=vs-2019#multidimensional-properties) to the list.

```
public class Example
{
  public bool Carrying[Robot r, Ball b]
  {
    get { return Database.Query(r, b); }
    set { Database.Store(r, b, value); }
  }
}
```

.NET and C++/CLI support multidimensional properties. Multidimensional properties can occur in .NET assemblies. It appears that C# doesn’t yet support the scenario – one can’t yet utilize multidimensional properties (other than indexers) in C#. See also: issue https://github.com/dotnet/csharplang/issues/471 where "multidimensional properties" are referred to as "indexed properties".

P.S.: The LINQ expression tree model and the dynamic language runtime could also be relevant in the context of multidimensional properties or indexed properties. Developers would expect to be able to utilize any new C# syntax with the `dynamic` keyword, in LINQ expressions, and in lambda expressions.

**Comment by CyrusNajmabadi** [comment]
> Predicates, numeric fluents and object fluents more closely resemble properties than either methods or delegates as one can both get values from and assign values to compound expressions.

You can do that with both delegates and methods. :)  

> `P(x, y, z) = true;`

this is already supported in C# today.

**Comment by AdamSobieski** [comment]
That is interesting that this can be achieved in C# today. Could you provide an example, hyperlink or gist?

**Comment by CyrusNajmabadi** [comment]
Sure!

```c#
ref bool P(int x, int y, int z) { ... }

// ...

P(1, 2, 3) = true;
```

**Comment by CyrusNajmabadi** [comment]
Alternatively, just use indexers, and the user can write:

```c#
P[1, 2, 3] = true;
```

Supported since C# 1.0.  Nice and simple and easy to understand.

**Comment by AdamSobieski** [comment]
Thank you. It looks like one can achieve predicates, numeric fluents and object fluents in various ways in C#.

In addition to the "()" brackets, I also like the "[]" brackets. With "[]" brackets, source code might be more readable with simultaneous uses of predicates and methods.

A reason that multidimensional or indexed properties could be useful for predicates involves storing them in assemblies. Similar to extension methods, one of a number of ways that a set of predicates could be stored in an assembly for use across .NET languages resembles:

```
namespace Namespace
{
  [Metadata]
  [CompilerGenerated]
  public static class Domain
  {
    [Predicate]
    public static bool Carrying[Robot r, Ball b]
    {
      [Metadata]
      get { return (bool)Database.Query("Namespace.Domain.Carrying", r, b); }
      [Metadata]
      set { Database.Store("Namespace.Domain.Carrying", r, b, value); }
    }
  }
}
```

Languages with support for first-class predicates could, then, find these metadata-described predicates in assemblies and provide them for developers.

Some metadata attributes would be useful for ensuring that static properties of static types were, indeed, predicates. Other metadata would be useful (e.g. upon getters and setters) for differentiating kinds of predicates or for describing their implementation details.

**Comment by CyrusNajmabadi** [comment]
> In addition to the "()" brackets, I also like the "[]" brackets. With "[]" brackets, source code might be more readable with simultaneous uses of predicates and methods.

Note, this is possible today in C#, just slightly annoying for the library author (not hte consumer).  Today, you can write:

```c#
  public static class Domain
  {
    public static CarryingThingy Carrying => new CarryingThingy(this);

    public struct CarryingThingy
    {
        private readonly Domain _domain;
        internal CarryingThingy(Domain domain) => _domain = domain;

        public bool this[Robot r, Ball b]
        {
            get { return (bool)_domain.Database.Query("Namespace.Domain.Carrying", r, b); }
            set { _domain.Database.Store("Namespace.Domain.Carrying", r, b, value); }
        }
     }
  }
```

consumer can then write: `myDomain.Carrying[robby, bally] = true`

**Comment by CyrusNajmabadi** [comment]
If you're generating this (which seems to be the case with `[CompilerGenerated]`) you could certainly generate those structs with the exposed indexers.

Note: this approach has practically no overhead at runtime.  .net core is great at just eliding all of this.  so `myDomain.Carrying[robby, bally] = true` basically is rewritten directly to `myDomain.Database.Store(...)`.  

This approach of typesafe struct wrappers is *super* common and performant in .net.

**Comment by CyrusNajmabadi** [comment]
> Some metadata attributes would be useful for ensuring that static properties of static types were, indeed, predicates

This feels like this could be done by you using Roslyn Analyzers: https://docs.microsoft.com/en-us/visualstudio/code-quality/roslyn-analyzers-overview?view=vs-2019

You can define your own attributes and encode whatever semantics you want for them.  you tehn write an analyzer to catch misuses of them. :)

**Comment by AdamSobieski** [comment]
Thanks again. There are now multiple possibilities for storing sets of predicates, numeric fluents and object fluents in .NET assemblies.

Interestingly, logical rules could be added to `Database` instances. Queries of `Database` instances could, then, return stored as well as derived values for predicates, numeric fluents and object fluents.

See also: [Datalog](https://en.wikipedia.org/wiki/Datalog), [Prolog](https://en.wikipedia.org/wiki/Prolog)

**Comment by AdamSobieski** [comment]
The [using static directive](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-static) is interesting with both approaches.

With the approach that @CyrusNajmabadi shared, one can explore first-class predicates in C# today:

```cs
using System;
using System.Runtime.CompilerServices;
using static Namespace.Domain;

namespace Namespace
{
    public interface IKnowledgebase
    {
        object Query(string predicate, params object[] tuple);
        void Store(string predicate, object value, params object[] tuple);
        /* ... */
    }

    public class Domain
    {
        public class Robot { }
        public class Ball { }

        static IKnowledgebase s_storage = null;

        public static _CarryingPredicate Carrying
        {
            get
            {
                return new _CarryingPredicate(s_storage);
            }
        }

        public struct _CarryingPredicate
        {
            public _CarryingPredicate(IKnowledgebase kb)
            {
                this.m_kb = kb;
            }
            readonly IKnowledgebase m_kb;

            public bool this[Robot r, Ball b]
            {
                [MethodImpl(MethodImplOptions.AggressiveInlining)]
                get { return (bool)m_kb.Query("Namespace.Domain.Carrying", r, b); }
                [MethodImpl(MethodImplOptions.AggressiveInlining)]
                set { m_kb.Store("Namespace.Domain.Carrying", value, r, b); }
            }
        }
    }

    public class Consumer
    {
        public void foo(Robot robot, Ball ball)
        {
            if (Carrying[robot, ball]) // first-class predicates
            {

            }
        }
    }
}
```

**Comment by AdamSobieski** [comment]
For those interested in these topics, I've started some new sketches on these topics which are available here: https://github.com/w3c-cg/planning/discussions/28 . The new sketches are of a C/C++/C#/Java/JavaScript-inspired language with more first-class language constructs including: templates, types, operators, predicates, rules, expressions, graphs, constraints, preferences, and attributes/decorators.

The language being sketched [there](https://github.com/w3c-cg/planning/discussions/28) isn't one intended for being compiled to a DLL or JAR, or to assembly code. It is, instead, intended for being [transpiled](https://en.wikipedia.org/wiki/Source-to-source_compiler) into a Semantic Web format, e.g., [TriG](https://www.w3.org/TR/trig/). Planners and solving algorithms, would, as envisioned, load input resources in this format (e.g., from a compressed archive file and/or from a number of Web resources).

Interestingly, transpiling allows first-class attributes and decorators to be able to update, to add to, the output graphs of other language constructs during the process of transpiling. This feature is introduced in the sketch available here: https://github.com/w3c-cg/planning/discussions/28#discussioncomment-4470596 . In the sketch, an example attribute or decorator is defined, `[comment("...")]`, which adds comment-related content to the Semantic Web outputs of other constructs during the transpilation process.

The language sketches are a work in progress and there remains much work to be done. Maybe these sketches can inspire some new C#/.NET features. Comments, feedback, questions, are welcomed either here or [there](https://github.com/w3c-cg/planning/discussions/28)!

**Reply by CyrusNajmabadi** [reply]
@AdamSobieski is there a suggestion you're making here for c#?  Thanks!

**Reply by AdamSobieski** [reply]
Perhaps [templates](https://en.wikipedia.org/wiki/Template_(C%2B%2B)) and [decorators](https://github.com/tc39/proposal-decorators) would be useful for C#/.NET?

Templates would allow genericity beyond that presently available with generic types, allowing template parameters and arguments beyond types. One could, for example, utilize a numeric type, an integer, as a template argument (`var t = new MyTemplate<123>();` or `var t = new MyTemplate<123>.NestedClass();`).

Decorators would resemble metadata attributes but would modify those things decorated, perhaps utilizing .NET's reflection and emitting capabilities. Perhaps a new thing, `System.Decorator`, could extend `System.Attribute`. As envisioned, decorators would be recognized and utilized by the content loading processes of the runtime, and a virtual method on them, e.g., `Process()`, would be invoked, receiving the decorated thing and outputting a modified thing.

**Reply by CyrusNajmabadi** [reply]
> Decorators

Can you give an example of something where this would be useful?  Thanks!

**Reply by AdamSobieski** [reply]
While .NET attributes can be added to many things, with respect to the usefulness of decorators, we can first consider decorating methods and decorating classes.

With respect to decorating methods, as envisioned, the overridden `Process()` method inherited from `System.Decorator` would receive a decoration context and a `MethodInfo` instance and would be expected to return another `MethodInfo`. For an example, a decorator might inject a call to a logger, or output something to the console, whenever the decorated method was invoked.

```cs
public class MyMethodDecorator : System.Decorator
{
    public MyMethodDecorator(string text)
    {
        m_text = text;
    }

    private string m_text;

    public override object Process(DecorationContext ctx, object that)
    {
        MethodInfo method = that as MethodInfo;
        MethodBuilder builder = ctx.CreateMethodBuilder(method);
        // ... 
        return builder;
    }
}

public class Example
{
    [MyMethodDecorator("Hello world.")]
    public void foo()
    {
    
    }
}
```

The above example resembles concepts from [aspect-oriented programming](https://en.wikipedia.org/wiki/Aspect-oriented_programming). The example decorator injects, prefixes, the decorated method's implementation with some logging, tracing, or debugging functionality.

Similarly, one could decorate an entire class, similarly reflecting on the decorated `Type` and emitting a modified `Type`.

```cs
public class MyClassDecorator : System.Decorator
{
    public MyClassDecorator() { }

    public override object Process(DecorationContext ctx, object that)
    {
        Type type = that as Type;
        TypeBuilder builder = ctx.CreateTypeBuilder(type);
        // ...
        return builder;
    }
}

[MyClassDecorator]
public class Example
{
    public void foo()
    {
    
    }
}
```

What do you think?

**Reply by HaloFour** [reply]
Source generators cover the use cases of using metadata like attributes to generate additional source.  However, the team has rejected language enhancements that would have enabled it for AOP scenarios.

See: #107

Although your examples don't seem like they would have anything to do with C# as a language and would instead be a post-compilation step, like [PostSharp](https://www.postsharp.net/aop.net).  Or as a part of the runtime itself.

**Reply by CyrusNajmabadi** [reply]
As this was linked from another thread, i wanted to add the same feedback as Halo mentioned.  This doesn't seem to be part of C#.  And it seems to me to be better served by a dedicated library for "Planning Domain Modeling".  If actual language features are needed here, a strong argument has to be provided as to why C# should now be aware of this esoteric domain, instead of someone interested in this esoteric domain simply providing a useful API with the right concepts that someone would interact with instead.
